File: nexus/src/main/java/io/xj/nexus/model/SegmentMeme.java
// Copyright (c) 1999-2022, XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

package io.xj.nexus.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Objects;
import java.util.UUID;


public class SegmentMeme {

  UUID id;
  UUID segmentId;
  String name;

  /**
   *
   **/
  public SegmentMeme id(UUID id) {
    this.id = id;
    return this;
  }


  @JsonProperty("id")
  public UUID getId() {
    return id;
  }

  public void setId(UUID id) {
    this.id = id;
  }

  /**
   *
   **/
  public SegmentMeme segmentId(UUID segmentId) {
    this.segmentId = segmentId;
    return this;
  }


  @JsonProperty("segmentId")
  public UUID getSegmentId() {
    return segmentId;
  }

  public void setSegmentId(UUID segmentId) {
    this.segmentId = segmentId;
  }

  /**
   *
   **/
  public SegmentMeme name(String name) {
    this.name = name;
    return this;
  }


  @JsonProperty("name")
  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }


  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    SegmentMeme segmentMeme = (SegmentMeme) o;
    return Objects.equals(this.id, segmentMeme.id) &&
      Objects.equals(this.segmentId, segmentMeme.segmentId) &&
      Objects.equals(this.name, segmentMeme.name);
  }

  @Override
  public int hashCode() {
    return Objects.hash(id, segmentId, name);
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("class SegmentMeme {\n");

    sb.append("    id: ").append(toIndentedString(id)).append("\n");
    sb.append("    segmentId: ").append(toIndentedString(segmentId)).append("\n");
    sb.append("    name: ").append(toIndentedString(name)).append("\n");
    sb.append("}");
    return sb.toString();
  }

  /**
   * Convert the given object to string with each line indented by 4 spaces
   * (except the first line).
   */
  String toIndentedString(Object o) {
    if (o == null) {
      return "null";
    }
    return o.toString().replace("\n", "\n    ");
  }
}

File: engine/src/segment/SegmentMeme.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include "xjmusic/segment/SegmentMeme.h"

using namespace XJ;


bool SegmentMeme::equals(const SegmentMeme &segmentMeme) const {
  return id == segmentMeme.id &&
         segmentId == segmentMeme.segmentId &&
         name == segmentMeme.name;
}


unsigned long long SegmentMeme::hashCode() const {
  return std::hash<std::string>{}(id) ^
         std::hash<int>{}(segmentId) ^
         std::hash<std::string>{}(name);
}


std::set<std::string> SegmentMeme::getNames(const std::set<SegmentMeme *> &segmentMemes) {
  std::set<std::string> names;
  for (const auto &segmentMeme: segmentMemes) {
    names.insert(segmentMeme->name);
  }
  return names;
}

